evennia/server/profiling/profiler.py

```diff
@@ -0,0 +1,177 @@
+import gc
+import inspect
+import tracemalloc
+import linecache
+import sys
+from datetime import datetime
+from functools import wraps
+
+TRASH = [
+    tracemalloc.Filter(False, "<frozen importlib._bootstrap>"),
+    tracemalloc.Filter(False, "<frozen importlib._bootstrap_external>"),
+    tracemalloc.Filter(False, "<frozen ntpath>"),
+    tracemalloc.Filter(False, "<frozen abc>")
+]
+
+KB = 1024
+MB = 1024*1024
+GB = 1024*1024*1024
+
+# snaps = []
+
+# def snapshot():
+#     snaps.append(tracemalloc.take_snapshot())
+
+# def trace(func, stream=None, frames=None):
+#     if stream is None:
+#         stream = sys.stdout
+
+#     @wraps(func)
+#     def wrapper(*args, **kwargs):
+#         if not tracemalloc.is_tracing():
+#             if frames:
+#                 tracemalloc.start(frames)
+#             else:
+#                 tracemalloc.start()
+
+#         result = func(*args, **kwargs)
+
+#         collect()
+#         snapshot()
+#         tracemalloc.stop()
+#         top(stream)
+
+#         return result
+
+#     return wrapper
+
+def top(snapshot, key_type='lineno', limit=10, precision=1, filters=[], stream=None):
+    if stream is None:
+        stream = sys.stdout
+
+    frame = inspect.currentframe().f_back
+    info = inspect.getframeinfo(frame)
+    file = info.filename
+    lineno = frame.f_lineno
+
+    # always filter trash
+    snapshot = snapshot.filter_traces(TRASH)
+
+    if len(filters):
+        if "filename" in filters:
+            snapshot = snapshot.filter_traces([
+                tracemalloc.Filter(True, filename_pattern=file)
+            ])
+        else:
+            snapshot = snapshot.filter_traces(filters)
+
+    stats = snapshot.statistics(key_type)
+
+
+    stream.write("[{}] <{}:{}>\n".format(datetime.now(), file, lineno))
+    stream.write("====[ EVENNIA TRACE ]====\n")
+    for index, stat in enumerate(stats[:limit], 1):
+        tb = stat.traceback[0]
+        stream.write("#{0}: {1}:{2} {3:.{4}f} KB\n".format(index, tb.filename, tb.lineno, stat.size / KB, precision))
+        line = linecache.getline(tb.filename, tb.lineno).strip()
+        if line:
+            stream.write("\t{}\n".format(line))
+    stream.write("\n")
+
+    remaining = stats[limit:]
+    if remaining:
+        size = sum(stat.size for stat in remaining)
+        stream.write("Output truncated at {} lines\n".format(limit))
+        stream.write("{} remaining lines: {} KB\n".format(len(remaining), size / 1024))
+    
+    total = sum(stat.size for stat in stats)
+    stream.write("Total allocated size: {} KB\n".format(total / 1024))
+    stream.write("\n")
+
+def compare(start, end, limit=10, stream=None):
```

Comment:
Docstrings are required for all functions and methods. We don't use Python typehints at this time, so adhering to Google codestyle type descriptors is paramount.